phpBB/phpbb/storage/controller/file.php

```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\storage\controller;
+
+use phpbb\config;
+use phpbb\db\driver\driver_interface;
+use phpbb\exception\http_exception;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+
+// /download/file.php
+
+class file
+{
+	/**
+	 * @var \phpbb\config\config
+	 */
+	protected $config;
+
+	/**
+	* @var ContainerInterface
+	*/
+	protected $container;
+
+	/**
+	 * @var \phpbb\db\driver\driver_interface
+	 */
+	protected $db;
+
+	protected $attachments_table;
+
+	/**
+	* Constructor.
+	*
+	* @param ContainerInterface $container A ContainerInterface instance
+	*/
+	public function __construct(config $config, ContainerInterface $container, db $db, $attachments_table)
+	{
+		$this->config = $config;
+		$this->container = $container;
+		$this->db = $db;
+		$this->attachments_table = $attachments_table;
+	}
+
+	/**
+	 * Controller for /download/{type}/{id} routes
+	 *
+	 * Because of how phpBB organizes routes $mode must be set in the route config.
+	 *
+	 * @param int		$id		ID of the entity to report
+	 * @param string	$mode
+	 * @return \Symfony\Component\HttpFoundation\Response a Symfony response object
+	 * @throws \phpbb\exception\http_exception when $mode or $id is invalid for some reason
+	 */
+	public function handle($type, $id)
+	{
+		$storage = $this->container->get('storage.'.$type);
+
+		switch($type)
+		{
+			case 'attachment':
+				$file = $this->get_attachment_file($id);
+				break;
+			case 'avatar':
+				$file = $this->get_avatar_file($id);
+				break;
+		}
+
+		echo $storage->get_contents($file)
```

Comment:
return new BinaryResponse($storage->get_url($file)); or something like that (it's a method to add. for cloud storage it may return URL with vendor prefix like aws:// per example)